miden-tx/src/host/tx_authenticator.rs

```diff
@@ -0,0 +1,158 @@
+    ///   authenticator to review any alterations to the account prior to signing.
+    ///   It should not be directly used in the signature computation.
+    fn get_signature(
+        &mut self,
+        pub_key: Word,
+        message: Word,
+        account_delta: &AccountDelta,
+    ) -> Result<Vec<Felt>, AuthenticationError>;
+}
+
+// BASIC AUTHENTICATOR
+// ================================================================================================
+
+#[derive(Clone, Debug)]
+pub enum KeySecret {
+    RpoFalcon512(rpo_falcon512::SecretKey),
+}
+
+#[derive(Clone, Debug)]
+/// Represents a signer for [KeySecret] keys
+pub struct BasicAuthenticator<R> {
+    /// pub_key |-> secret_key mapping
+    keys: BTreeMap<Digest, KeySecret>,
+    rng: R,
+}
+
+impl<R: Rng> BasicAuthenticator<R> {
+    #[cfg(feature = "std")]
+    pub fn new(keys: &[(Word, KeySecret)]) -> BasicAuthenticator<rand::rngs::StdRng> {
+        use rand::{rngs::StdRng, SeedableRng};
+
+        let rng = StdRng::from_entropy();
+        BasicAuthenticator::<StdRng>::new_with_rng(keys, rng)
+    }
+
+    pub fn new_with_rng(keys: &[(Word, KeySecret)], rng: R) -> Self {
+        let mut key_map = BTreeMap::new();
+        for (word, secret_key) in keys {
+            key_map.insert(word.into(), secret_key.clone());
+        }
+
+        BasicAuthenticator { keys: key_map, rng }
+    }
+}
+
+impl<R: Rng> TransactionAuthenticator for BasicAuthenticator<R> {
+    /// Gets as input a [Word] containing a secret key, and a [Word] representing a message and
+    /// outputs a vector of values to be pushed onto the advice stack.
+    /// The values are the ones required for a Falcon signature verification inside the VM and they are:
+    ///
+    /// 1. The nonce represented as 8 field elements.
+    /// 2. The expanded public key represented as the coefficients of a polynomial of degree < 512.
+    /// 3. The signature represented as the coefficients of a polynomial of degree < 512.
+    /// 4. The product of the above two polynomials in the ring of polynomials with coefficients
+    /// in the Miden field.
+    ///
+    /// # Errors
+    /// Will return an error if either:
+    /// - The secret key is malformed due to either incorrect length or failed decoding.
+    /// - The signature generation failed.
+    fn get_signature(
+        &mut self,
+        pub_key: Word,
+        message: Word,
+        account_delta: &AccountDelta,
+    ) -> Result<Vec<Felt>, AuthenticationError> {
+        let _ = account_delta;
+
+        let secret_key = match self.keys.get(&pub_key.into()) {
+            Some(key) => match key {
+                KeySecret::RpoFalcon512(falcon_key) => falcon_key,
+            },
+            None => {
+                return Err(AuthenticationError::UnknownKey(format!(
+                    "Public key {} is not contained in the authenticator's keys",
+                    Digest::from(pub_key)
+                )))
+            },
+        };
```

Comment:
I would probably keep just the dispatch in this method and move the Falcon-specific code into a helper functions - e.g., something like:
```Rust
fn get_falcon_signature<R: Rng>(
    key: rpo_falcon512::SecrtKey,
    message: Word,
    rng: &mut R,
) -> Result<Vec<Felt>, AuthenticationError> {
    tood!()
}
```

Reply:
Changed